docs/design/values.md

```diff
@@ -0,0 +1,917 @@
+can appear where a member declaration would be valid within the type:
+
+```carbon
+class SomeType {
+  value_rep = RepresentationType;
+}
+```
+
+**Open question:** The syntax for this is just placeholder, using a placeholder
+keyword. It isn't final at all and likely will need to change to read well.
+
+The provided representation type must be one of the following:
+
+-   `const Self` -- this forces the use of a _copy_ of the object.
+-   `const Self *` -- this forces the use of a [_pointer_](#pointers) to the
+    original object.
+-   A custom type that is not `Self`.
+
+If the representation is `const Self` or `const Self *`, then the type fields
+will be accessible as [_value expressions_](#value-expressions) using the normal
+member access syntax for value expressions of a type. These will be implemented
+by either accessing a copy of the object in the non-pointer case or a pointer to
+the original object in the pointer case. A representation of `const Self`
+requires copying to be valid for the type. This provides the builtin
+functionality but allows explicitly controlling which representation should be
+used. If no customization is provided, the implementation will select one based
+on a set of heuristics.
+
+When a custom type is provided, it must not be `Self` (or the other two cases).
+The type provided will be used on function call boundaries and as the
+implementation representation for `let` bindings and other value expressions
+referencing an object of the type. A specifier of `value_rep = T;` will require
+that the type `impls ReferenceImplicitAs(T)` using the following interface:
+
+```carbon
+interface ReferenceImplicitAs(T:! type) {
+  fn Op[addr self: const Self*]() -> T;
+}
+```
+
+Converting a reference expression into a value expression for such a type calls
+this customization point to form a representation object from the original
+reference expression.
+
+When using a custom representation type in this way, no fields are accessible
+through a value expression. Instead, only methods can be called using member
+access, as they simply bind the value expression to the `self` parameter.
+However, one important method can be called -- `.(ImplicitAs(T).Convert()`. This
+implicitly converting a value expression for the type into its custom
+representation type. The customization of the representation above causes the
+class to have a builtin `impl as ImplicitAs(T)` which converts to the
+representation type as a no-op, exposing the object created by the
+`ReferenceImplicitAs` operation, and preserved as a representation of the value
+expression.
+
+Here is a more complete example of code using these features:
+
+```carbon
+class StringView {
+  // A typical readonly view of a string API...
+  fn ExampleMethod[self: Self]() { ... }
+}
+
+class String {
+  // Customize the value representation to be `StringView`.
+  value_rep = StringView;
+
+  private var data_ptr: Char*;
+  private var size: i64;
```

Comment:
In particular, it would be good to explain what happens with methods that would be `const` in C++, but can't be called with just the information available in the r-value representation.